enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,940 @@
+and provides information on what should admin do to retry upgrade.
+
+1. Commit with MicroShift 4.13 is running
+1. Admin prepares and stages new commit that includes newer MicroShift,
+   and reboots the device.
+1. MicroShift's data exists, but metadata with version is missing,
+   so it is assumed that upgrade is from 4.13.
+   - Back up 4.13 data
+   - If upgrade from 4.13 is supported, then try to migrate the data.
+     Otherwise refuse (renders system unhealthy and causes a rollback).
+1. MicroShift starts the cluster
+1. System ends up unhealthy for any reason. It could be that:
+   - Upgrade was blocked or storage migration failed
+   - MicroShift was unhealthy
+   - Something else was unhealthy
+1. Device is rebooted by greenboot
+1. Because _previous boot_ was unhealthy,
+   there is no backup for _current boot's commit_,
+   _previous commit with MicroShift_ is the same as _rollback_:
+   restore backup of _previous commit with MicroShift_ and attempt to perform
+   the upgrade again.
+1. System happens to be consistently unhealthy and rolls back to commit with 4.13
+1. MicroShift cluster starts, we cannot predict if it fails or not
+1. Regardless if system is healthy or not, admin still wants to upgrade.
+1. Admin manually restores backup compatible with 4.13 and addresses everything
+   that might have affected status of upgrade.
+1. Admin prepares and stages another commit with newer MicroShift,
+   and reboots the device.
+1. MicroShift's data exists, but metadata with version is missing,
+   so it is assumed that upgrade is from 4.13.
+   - Back up 4.13 data
+   - If upgrade from 4.13 is supported, then try to migrate the data.
+     Otherwise refuse (renders system unhealthy and causes a rollback).
+
+#### Device runs pre-loaded commit with MicroShift
+
+Following section describes expected flows when device is running MicroShift
+for the first time.
+
+##### First boot of the pre-loaded commit with MicroShift
+
+1. Device is pre-loaded with commit including MicroShift and booted.
+1. MicroShift's data does not exist, so cluster is simply started.
+1. Greenboot performs health checks
+1. Depending on system's health
+   - System is healthy
+     1. [First commit, second boot (reboot): backup fails](#first-commit-second-boot-reboot-backup-fails)
+     1. [First commit, second boot (reboot): backup succeeds](#first-commit-second-boot-reboot-backup-succeeds)
+   - System is unhealthy
+     1. Greenboot doesn't reboot device (`boot_counter` is only set when ostree commit is staged)
+     1. System requires manual intervention.
+
+##### Device is rebooted (non-greenboot) and it fails to back up the data
+
+Continuation of [First boot of the pre-loaded commit with MicroShift](#first-boot-of-the-pre-loaded-commit-with-microshift)
+with assumption that system was healthy before the reboot.
+
+1. Devices is rebooted and boots the same commit
+1. Previous boot was healthy, but attempt to back up the data fails,
```

Comment:
does this clobber the previous backup data?

Reply:
Do you mean if by attempting to `cp` to the same target directory we might end up only partially copying files and effectively corrupting the existing healthy backup?
That's a valid scenario, I think that we should try to `cp` to temporary dir (sibling to existing backup) and then do a rename (sort of "atomic copy")

Reply:
> Do you mean if by attempting to cp to the same target directory we might end up only partially copying files and effectively corrupting the existing healthy backup?

exactly

> That's a valid scenario, I think that we should try to cp to temporary dir (sibling to existing backup) and then do a rename (sort of "atomic copy")

yeah that seems reasonable

Reply:
included info about cp & rename